### PR TITLE
pkg/urlutil: deprecate, and move to builder/remotecontext/urlutil

### PR DESCRIPTION
**Pull request by thaJeztah**
follow-up to https://github.com/moby/moby/pull/43476 (only last three commits are new)

### pkg/urlutil: remove unused IsTransportURL()

This function is no longer used (either internally, or externally), so can be removed.

### pkg/urlutil: deprecate, and move to builder/remotecontext/urlutil

pkg/urlutil (despite its poorly chosen name) is not really intended as a generic utility to handle URLs, and should only be used by the builder to handle (remote) build contexts.

- IsURL() only does a very rudimentary check for http(s):// prefixes, without any other validation, but due to its name may give incorrect expectations.
- IsGitURL() is written specifically with docker build remote git contexts in mind, and has handling for backward-compatibility, where strings that are not URLs, but start with "github.com/" are accepted.

Because of the above, this patch:

- moves the package inside builder/remotecontext, close to where it's intended to be used (ideally this would be part of build/remotecontext itself, but this package imports many other dependencies, which would introduce those as extra dependencies in the CLI).
- deprecates pkg/urlutil, but adds aliases as there are some external consumers.

### builder/remotecontext/urlutil: simplify and improve documentation

Simplify some of the logic, and add documentation about the package, as well as warnings that this package should not be used as a general-purpose utility.

**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

**Comment by thaJeztah** [comment]
@tonistiigi @crazy-max ptal

File: builder/remotecontext/urlutil/urlutil.go [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+// "container" in the https://github.com/myorg/my-repo.git repository:
+//
+// https://github.com/myorg/my-repo.git#container:docker
+//
+// The following table represents all the valid suffixes with their build
+// contexts:
+//
+// | Build Syntax Suffix            | Git reference used   | Build Context Used |
+// |--------------------------------|----------------------|--------------------|
+// | my-repo.git                    | refs/heads/master    | /                  |
```
Comment (thaJeztah) [inline builder/remotecontext/urlutil/urlutil.go:70]: let me change this to `main` instead of `master` (and describe "default branch")

Comment (thaJeztah) [reply builder/remotecontext/urlutil/urlutil.go:70]: oh, actually, looks like we still have that hardcoded, so we'll have to look at that in a follow-up.

**Comment by thaJeztah** [comment]
Ah dang; can't create an alias;

```
These files import internal code: (either directly or indirectly)
  - pkg/urlutil/deprecated.go imports github.com/docker/docker/builder/remotecontext/urlutil

File: pkg/urlutil/deprecated.go [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+
+import "github.com/docker/docker/builder/remotecontext/urlutil"
+
+var (
+	// IsURL returns true if the provided str is an HTTP(S) URL.
+	//
+	// Deprecated: use github.com/docker/docker/builder/remotecontext/urlutil.IsURL
+	// to detect build-context type, or use strings.HasPrefix() to check if the
+	// string has a https:// or http:// prefix.
+	IsURL = urlutil.IsURL
```
[the remark this answers is not in the document]
Comment (thaJeztah) [reply pkg/urlutil/deprecated.go:12]: Bit of a habit (not sure if "good" or "bad") it's slightly easier to create a list of aliases when deprecating a package (which _could_ be slightly clearer "hey, don't use this; here's where to look"), and in some cases may remove imports that are used in the signature of the aliased function 😅

[the remark this answers is not in the document]
Comment (thaJeztah) [reply pkg/urlutil/deprecated.go:12]: Yup, you're right. In this case it doesn't bring much; let me change it

[the remark this answers is not in the document]
Comment (thaJeztah) [reply pkg/urlutil/deprecated.go:12]: Updated 👍

**Review by tonistiigi — APPROVED** [review]
I wouldn't bother with `pkg/urlutil` and just remove it.

**Comment by thaJeztah** [comment]
Yeah, definitely not planning to keep it for long, but found it used in various external projects, so wanted to give a tiny amount of room for those to move away from it.

**Comment by thaJeztah** [comment]
Interesting error on CI (DCO check test)

```bash
git openssh-client && cd /workspace && hack/validate/dco
[2022-04-12T21:37:15.034Z] Unable to find image 'alpine:latest' locally
[2022-04-12T21:37:15.293Z] latest: Pulling from library/alpine
[2022-04-12T21:37:15.293Z] df9b9388f04a: Pulling fs layer
[2022-04-12T21:37:15.860Z] df9b9388f04a: Pull complete
[2022-04-12T21:37:15.860Z] Digest: sha256:4edbd2beb5f78b1014028f4fbb99f3237d9561100b6881aabbf5acce2c4f9454
[2022-04-12T21:37:15.860Z] Status: Downloaded newer image for alpine:latest
[2022-04-12T21:37:23.999Z] fatal: unsafe repository ('/workspace' is owned by someone else)
[2022-04-12T21:37:23.999Z] To add an exception for this directory, call:
[2022-04-12T21:37:23.999Z] 
[2022-04-12T21:37:23.999Z] 	git config --global --add safe.directory /workspace
script returned exit code 128
```

**Comment by thaJeztah** [comment]
I see it's using an `alpine:latest` image to run the check in; wondering if alpine was updated (or the git version in alpine)

**Review by samuelkarp — APPROVED** [review]
LGTM aside from whatever weirdness is going on with CI.

**Comment by thaJeztah** [comment]
Looks to be related to a vulnerability that was fixed in git; https://github.blog/2022-04-12-git-security-vulnerability-announced/

**Comment by thaJeztah** [comment]
Triggered CI with the DCO check skipped (to check if the rest of the pipeline also needs adjustments)

I'll have a look at fixing the DCO stage to adjust for the git security fix (looks to be because the source is checked out on the worker by Jenkins, but then accessed inside a container, in which case the user doesn't match the user on the host, triggering the security warning)

**Pull request merged**